Refactor SubstituteContext to decouple the thread local context

zvirja opened the pull request:
Should be a huge step forward towards the #150.

First of all please take my apology for the HUGE PR. Let me know if you wish me to split it into the chunks. I decided to push all the changes together because for me they look logically coupled and I don't see too much benefit from pushing changes one by one. Probably, it's easier to review PR commit-by-commit 😅

------
I've found that currently `SubstitutionContext` depends on `SubstituteFactory`, while `SubstituteFactory` depends on the `SubstitutionContext` to access the thread-local data, leading to the circular dependency. It makes it complicated to pass the customized `SubstituteFactory`, while [we offer](https://github.com/nsubstitute/NSubstitute/blob/e63e73dbcaccd726f57750617059041296683c57/src/NSubstitute/Core/SubstitutionContext.cs#L41) such an API. Another observation was that currently `SubstitutionContext` carries too much responsibilities by holding the root factories/resolvers and offering the thread-local storage in the same time.

To improve the situation and make the overall code much nicer and transparent I applied the following changes:
- Extracted all the thread-local stuff to the `IThreadLocalContext`. Later I fixed all the places in the code to depend on the `IThreadLocalContext` rather than on `ISubstitutionContext`. It made it clear what exactly we use in the particular class.
- Obsoleted all the members in the `ISubstitutionContext` related to work with thread-local data. This API should be preserved at least for a couple of next major releases and a lot of libraries might depend on that API (e.g. AutoFixture). That's especially true for `GetRouteFactory()`.
- Cleaned up the `ISubstituteState` interface and implementation to not carry the thread-local data, as it isn't a part of the particular substitute state. It makes access to the thread-local data more explicit. Surprisingly, it required a few minor tweaks only to achieve that goal:
    - Move `IPendingSpecification` directly to the `IThreadContext`
    - Adjust the `IConfigureCall` and `IGetCallSpec` to accept pending specification info as a parameter, rather than work with `IPendingSpecification` internally. 
    - Pass `IThreadLocalContext` to the `RouteFactory` constructor, so this context is global to the factory, rather than specific to the particular substitute. In fact that's always true for our existing architecture and is one of its particularity. Therefore, I find this change fine.
- I found that currently last call result is always configured using the `IThreadLocalContext.LastCallShouldReturn()` and there we apply most of the checks (e.g. last call router is known; no pending specification is present). Therefore, I removed the work with the `IPendingSpecification` from the `CallRouter` and dequeue pending specification directly in the `ThreadLocalContext` class. As result, the `IsLastCallInfoPresent()` method has been removed and the `LastCallShouldReturn()` method receives the `PendingSpecInfo` now. Honestly, I have some doubts regarding the change, however the previous logic, when we had `ThreadLocalContext.LastCallShouldReturn() -> CallRouter.IsLastCallInfoPresent() -> ThreadLocalContext.PendingSpecification.HasPendingCallSpecInfo()` call chain looks a bit complicated. I have a few different views on this change (basically encapsulation vs complexity), so wouldn't mind any way.
- Use the single instance of `RouteFactory` everywhere, as the factory is stateless. That's required also because it captures a reference to the `IThreadLocalContext` instance now.
- Obsoleted the `SequenceNumberGenerator` property of the `ISubstitutionContext`, as it isn't referenced directly now. Rather, it's passed to the `CallRouterFactory` constructor and is captured there.
- Moved `ICallRouterResolver` to the `ISubstitutionContext` class as currently it isn't bound to the factory. let me know if you wish me to revert the change. Probably, the previous API looked more consistent, as in theory only factory knows how to recover the ICallRouter later... 😕 
-----------------------

I do understand that my PR might cause a ton of work to review and realize the changes I would like to apply, so I'll definitely owe you some beer afterwards 🍺😊 However, I feel that change should positively improve the existing architecture by making it more explicit.

@dtchepak @alexandrnikitin Please let me know your opinion regarding the suggested change.

dtchepak commented on src/NSubstitute/Core/SubstitutionContext.cs:
```diff
@@ -14,123 +12,118 @@ public class SubstitutionContext : ISubstitutionContext
+
+
+        // ***********************************************************
+        // ********************** OBSOLETE API **********************
```
Should these members be marked `[Obsolete]` then?

zvirja replied:
@dtchepak Nice catch, thanks! 👍 Due to some reasons I missed that I should mark implemented members as well :-/

dtchepak commented on src/NSubstitute/Core/ThreadLocalContext.cs:
```diff
@@ -41,16 +41,21 @@ public void SetLastCallRouter(ICallRouter callRouter)
             var lastCallRouter = _lastCallRouter.Value;
-            if (lastCallRouter == null) throw new CouldNotSetReturnDueToNoLastCallException();
-            if (!lastCallRouter.IsLastCallInfoPresent()) throw new CouldNotSetReturnDueToMissingInfoAboutLastCallException();
+            if (lastCallRouter == null)
```
Don't worry about changing it here, but for future changes I'd prefer to put braces around any `if` blocks that span multiple lines (I'm happy to omit them when they're on a single line as a guard).

e.g.

```
// Single line
if (condition) someCall();
// Multi line
if (condition) 
{
  someCall();
}

// Error prone:
if (condition)
   someCall();
   anotherCallThatLooksLikeItIsPartOfIfButIsNotDueToIndentingFail();
```

Happy to be convinced otherwise. :)

Aside: I actually prefer:
```
if (condition) {
  someCall();
}
```

But unfortunately that doesn't seem standard for C# :(

zvirja replied:
@dtchepak Roger that! 😅

zvirja commented:
@dtchepak I've addressed your concerns and marked the obsolete members in `SubstitutionContext` class as well.

Before you merge, I'd like to hear your opinion regarding the decision mentioned above:
> Moved ICallRouterResolver to the ISubstitutionContext class as currently it isn't bound to the factory. let me know if you wish me to revert the change. Probably, the previous API looked more consistent, as in theory only factory knows how to recover the ICallRouter later... 😕

How do you see it? Is it better to have `substitutionContext.GetCallRouterFor(subs)` or `substitutionContext.SubsituteFactory.GetCallRouterFor(subs)`? From one hand, it's the factory, that knows how to recover `ICallRouter` from the created substitutes (because I know how Castle works). From other hand, it might be non-obvious why you need to use factory and it looks like an extra responsibility.

The more I think, the more I want to make `GetCallRouterFor()` a part of the factory (revert the change I applied here). So, the `SubstitutionContext` contains the bare minimum... 😉

dtchepak commented:
I think I prefer `GetCallRouterFor` in the `SubstitutionContext`, I thought it was a good change! :) But I don't feel strongly about it so if you really want to move it then you are welcome to. :)

dtchepak commented:
@zvirja I'm happy with this PR. If you want to make more changes I'm happy to re-review, otherwise if you are happy enough as is please merge it in. :)

zvirja commented:
@dtchepak After playing more with API, I finally decided to keep `GetCallRouterFor()` as a part of the `SubstitutionContext`. If we move it under the `SubstituteFactory`, the usage will look like following:

```c#
var router = SubstitutionContext.Current.SubstituteFactory.GetCallRouterFor(subs);
```

It means that in order to retrieve the call router, you should always use the factory that activated that substitute. As a consequence, the following snippet of code is formally invalid:
```c#
var customFactory = new SubstituteFactory(... customized members ...);
var substitute = customFactory.Create(typeof(IInterface));
substitute.When(x => x.Call()).Do(_ => { callback });
```

The `When()` method will use the `SubstitutionContext.Current.SubstituteFactory` to resolve the call router, which should not work, as the different instance of the factory is used. The reason why it actually works, is the implementation detail of the factory.

Therefore, to make this scenario formally valid, I decided to put router resolver to `SubsitutionContext`. Now we say "you might use any factory to create a substitute - we'll use the same resolver to extract call router from it". It's more aligned with what we want to offer.
